lib/spack/spack/environment.py

```diff
@@ -1288,14 +1418,18 @@ def install_all(self, args=None):
         # a large amount of time due to repeatedly acquiring and releasing
         # locks, this does an initial check across all specs within a single
         # DB read transaction to reduce time spent in this case.
-        uninstalled_specs = []
+        specs_to_install = []
         with spack.store.db.read_transaction():
             for concretized_hash in self.concretized_order:
                 spec = self.specs_by_hash[concretized_hash]
-                if not spec.package.installed:
-                    uninstalled_specs.append(spec)
-
-        for spec in uninstalled_specs:
+                if not spec.package.installed or (
+                        spec.satisfies('dev_path=any') or
+                        spec.satisfies('^dev_path=any')
```

Comment:
This occurs in 2 places (well, one and one place that it could be extracted from an `any` block. IMO it's not worth changing. If I were to change it, I think `depends_on_develop` is too specific. We could add a `Spec.satisfies_or_dep_satisfies` method, but I don't think it's warranted yet at this point.